# Introduction

In this notebook, we will do a comprehensive analysis of the Android app market by comparing thousands of apps in the Google Play store.

# About the Dataset of Google Play Store Apps & Reviews

**Data Source:** <br>
App and review data was scraped from the Google Play Store by Lavanya Gupta in 2018. Original files listed [here](
https://www.kaggle.com/lava18/google-play-store-apps).

# Import Statements

In [ ]:
import pandas as pd
import plotly.express as px
import plotly.graph_objects as go
import matplotlib.pyplot as plt
import seaborn as sns

# Notebook Presentation

In [ ]:
# Show numeric output in decimal format e.g., 2.15
pd.options.display.float_format = '{:,.2f}'.format

# Read the Dataset

In [ ]:
df_apps = pd.read_csv('apps.csv')

# Data Cleaning

How many rows and columns does `df_apps` have? What are the column names? Look at a random sample of 5 different rows with [.sample()](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.sample.html).

In [ ]:
df_apps.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10841 entries, 0 to 10840
Data columns (total 12 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   App             10841 non-null  object 
 1   Category        10841 non-null  object 
 2   Rating          9367 non-null   float64
 3   Reviews         10841 non-null  int64  
 4   Size_MBs        10841 non-null  float64
 5   Installs        10841 non-null  object 
 6   Type            10840 non-null  object 
 7   Price           10841 non-null  object 
 8   Content_Rating  10841 non-null  object 
 9   Genres          10841 non-null  object 
 10  Last_Updated    10841 non-null  object 
 11  Android_Ver     10839 non-null  object 
dtypes: float64(2), int64(1), object(9)
memory usage: 1016.5+ KB


In [ ]:
df_apps.head()

App         Category  Rating  \
0                       Ak Parti Yardım Toplama           SOCIAL     NaN   
1                    Ain Arabic Kids Alif Ba ta           FAMILY     NaN   
2  Popsicle Launcher for Android P 9.0 launcher  PERSONALIZATION     NaN   
3                     Command & Conquer: Rivals           FAMILY     NaN   
4                                    CX Network         BUSINESS     NaN   

   Reviews  Size_MBs Installs  Type   Price Content_Rating           Genres  \
0        0      8.70        0  Paid  $13.99           Teen           Social   
1        0     33.00        0  Paid   $2.99       Everyone        Education   
2        0      5.50        0  Paid   $1.49       Everyone  Personalization   
3        0     19.00        0   NaN       0   Everyone 10+         Strategy   
4        0     10.00        0  Free       0       Everyone         Business   

     Last_Updated         Android_Ver  
0   July 28, 2017          4.1 and up  
1  April 15, 2016          3.0 and up  
2   July 11, 2018          4.2 and up  
3   June 28, 2018  Varies with device  
4  August 6, 2018          4.1 and up

We can already see that there are some data issues that we need to fix. In the Ratings and Type columns there are NaN (Not a number values) and in the Price column we have dollar signs that will cause problems.

In [ ]:
df_apps.sample(5)

App         Category  Rating  Reviews  Size_MBs  \
3761         BL Kennedy Theme  PERSONALIZATION    4.20      279      1.10   
4379            CS Guns Shoot           FAMILY    3.70      433     19.00   
5693                       FK          MEDICAL    4.40     1517      8.50   
3799  Oddworld: New 'n' Tasty             GAME    4.60     1475     34.00   
2138             DS-L4 Viewer      PHOTOGRAPHY    3.50       13      7.90   

     Installs  Type  Price Content_Rating           Genres      Last_Updated  \
3761   10,000  Free      0       Everyone  Personalization       May 6, 2013   
4379   50,000  Free      0           Teen    Entertainment    August 1, 2013   
5693  100,000  Free      0       Everyone          Medical     July 13, 2018   
3799   10,000  Paid  $4.99           Teen           Action   January 2, 2018   
2138    1,000  Free      0       Everyone      Photography  January 29, 2018   

     Android_Ver  
3761  2.1 and up  
4379  2.2 and up  
5693  2.3 and up  
3799  5.0 and up  
2138  5.0 and up

### Drop Unused Columns

Remove the columns called `Last_Updated` and `Android_Version` from the DataFrame. We will not use these columns.

In [ ]:
df_apps = df_apps.drop(columns=['Last_Updated', 'Android_Ver'])
df_apps.head()

App         Category  Rating  \
0                       Ak Parti Yardım Toplama           SOCIAL     NaN   
1                    Ain Arabic Kids Alif Ba ta           FAMILY     NaN   
2  Popsicle Launcher for Android P 9.0 launcher  PERSONALIZATION     NaN   
3                     Command & Conquer: Rivals           FAMILY     NaN   
4                                    CX Network         BUSINESS     NaN   

   Reviews  Size_MBs Installs  Type   Price Content_Rating           Genres  
0        0      8.70        0  Paid  $13.99           Teen           Social  
1        0     33.00        0  Paid   $2.99       Everyone        Education  
2        0      5.50        0  Paid   $1.49       Everyone  Personalization  
3        0     19.00        0   NaN       0   Everyone 10+         Strategy  
4        0     10.00        0  Free       0       Everyone         Business

### Find and Remove NaN values in Ratings

How may rows have a NaN value (not-a-number) in the Ratings column? Create DataFrame called `df_apps_clean` that does not include these rows.

In [ ]:
nan_ratings_rows = df_apps[df_apps['Rating'].isna()]
print(f'{len(nan_ratings_rows)} rows have NaN values in the Ratings column')

1474 rows have NaN values in the Ratings column


In [ ]:
df_apps_clean = df_apps.dropna(subset=['Rating'])

### Find and Remove Duplicates

**Challenge**: Are there any duplicates in data? Check for duplicates using the [.duplicated()](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.duplicated.html) function. How many entries can you find for the "Instagram" app? Use [.drop_duplicates()](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.drop_duplicates.html) to remove any duplicates from `df_apps_clean`.


In [ ]:
duplicate_rows = df_apps_clean[df_apps_clean.duplicated()]
print(f'{len(duplicate_rows)} rows are duplicates')

476 rows are duplicates


In [ ]:
df_apps_clean = df_apps_clean.drop_duplicates(subset=["App","Type","Price"])
df_apps_clean.shape

(8199, 10)

So we can see that 13 different features were originally scraped from the Google Play Store.



Obviously, the data is just a sample out of all the Android apps. It doesn't include all Android apps of which there are millions.

I’ll assume that the sample is representative of the App Store as a whole. This is not necessarily the case as, during the web scraping process, this sample was served up based on geographical location and user behaviour of the person who scraped it - in our case Lavanya Gupta.

The data was compiled around 2017/2018. The pricing data reflect the price in USD Dollars at the time of scraping. (developers can offer promotions and change their app’s pricing).

I’ve converted the app’s size to a floating-point number in MBs. If data was missing, it has been replaced by the average size for that category.

The installs are not the exact number of installs. If an app has 245,239 installs then Google will simply report an order of magnitude like 100,000+. I’ve removed the '+' and we’ll assume the exact number of installs in that column for simplicity.

# Find Highest Rated Apps

Identify which apps are the highest rated. What problem might you encounter if you rely exclusively on ratings alone to determine the quality of an app?

In [ ]:
df_apps_clean.sort_values(by="Rating", ascending=False).head(5)

App     Category  Rating  Reviews  \
21                         KBA-EZ Health Guide      MEDICAL    5.00        4   
1751                     Spine- The dating app       DATING    5.00        5   
1761             Selfie With Champion AJ Style  PHOTOGRAPHY    5.00        2   
1763                     Ag Valley Cooperative     BUSINESS    5.00        6   
1769  Yazdani Cd Center EllahAbad Official App       FAMILY    5.00        8   

      Size_MBs Installs  Type Price Content_Rating         Genres  
21       25.00        1  Free     0       Everyone        Medical  
1751      9.30      500  Free     0           Teen         Dating  
1761      7.50      500  Free     0       Everyone    Photography  
1763     74.00      500  Free     0       Everyone       Business  
1769      3.80      500  Free     0       Everyone  Entertainment

Only apps with very few reviews (and a low number on installs) have perfect 5 star ratings (most likely by friends and family).

# Find 5 Largest Apps in terms of Size (MBs)

What's the size in megabytes (MB) of the largest Android apps in the Google Play Store. Based on the data, do you think there could be limit in place or can developers make apps as large as they please?

In [ ]:
df_apps_clean.sort_values(by="Size_MBs", ascending=False).head(5)

App Category  Rating  Reviews  Size_MBs  \
8719          Draft Simulator for FUT 18   SPORTS    4.60   162933    100.00   
8718   Mini Golf King - Multiplayer Game     GAME    4.50   531458    100.00   
10295                    SimCity BuildIt   FAMILY    4.50  4218587    100.00   
9945                     Ultimate Tennis   SPORTS    4.30   183004    100.00   
9944        Gangster Town: Vice District   FAMILY    4.30    65146    100.00   

         Installs  Type Price Content_Rating      Genres  
8719    5,000,000  Free     0       Everyone      Sports  
8718    5,000,000  Free     0       Everyone      Sports  
10295  50,000,000  Free     0   Everyone 10+  Simulation  
9945   10,000,000  Free     0       Everyone      Sports  
9944   10,000,000  Free     0     Mature 17+  Simulation

Here we can clearly see that there seems to be an upper bound of 100 MB for the size of an app. A quick google search would also have revealed that this limit is imposed by the Google Play Store itself. It’s interesting to see that a number of apps actually hit that limit exactly.

# Find the 5 App with Most Reviews

Which apps have the highest number of reviews? Are there any paid apps among the top 50?

In [ ]:
df_apps_clean.sort_values(by="Reviews", ascending=False).head(50)

App            Category  \
10805                                           Facebook              SOCIAL   
10785                                 WhatsApp Messenger       COMMUNICATION   
10806                                          Instagram              SOCIAL   
10784           Messenger – Text and Video Chat for Free       COMMUNICATION   
10650                                     Clash of Clans                GAME   
10744            Clean Master- Space Cleaner & Antivirus               TOOLS   
10835                                     Subway Surfers                GAME   
10828                                            YouTube       VIDEO_PLAYERS   
10746  Security Master - Antivirus, VPN, AppLock, Boo...               TOOLS   
10584                                       Clash Royale                GAME   
10763                                   Candy Crush Saga                GAME   
10770        UC Browser - Fast Download Private & Secure       COMMUNICATION   
10735                                           Snapchat              SOCIAL   
10489    360 Security - Free Antivirus, Booster, Cleaner               TOOLS   
10731                                     My Talking Tom                GAME   
10594                                        8 Ball Pool                GAME   
10302  DU Battery Saver - Battery Charger & Battery Life               TOOLS   
10354                        BBM - Free Calls & Messages       COMMUNICATION   
10549  Cache Cleaner-DU Speed Booster (booster & clea...               TOOLS   
10757                                            Twitter  NEWS_AND_MAGAZINES   
10721                                    Viber Messenger       COMMUNICATION   
10578                                     Shadow Fight 2                GAME   
10813                                      Google Photos         PHOTOGRAPHY   
10724                        LINE: Free Calls & Messages       COMMUNICATION   
10717                                                Pou                GAME   
10792                      Skype - free IM & video calls       COMMUNICATION   
10628                                         Pokémon GO                GAME   
10388           Minion Rush: Despicable Me Official Game                GAME   
10694                                            Yes day                GAME   
10695                                            Hay Day              FAMILY   
10644                           Dream League Soccer 2018                GAME   
10696                                  My Talking Angela                GAME   
10660             VivaVideo - Video Editor & Photo Movie       VIDEO_PLAYERS   
10786                       Google Chrome: Fast & Secure       COMMUNICATION   
10817                          Maps - Navigate & Explore    TRAVEL_AND_LOCAL   
10672                                  Hill Climb Racing                GAME   
10734                                      Facebook Lite              SOCIAL   
10649                                Asphalt 8: Airborne                GAME   
10699                          Mobile Legends: Bang Bang                GAME   
10322  Battery Doctor-Battery Life Saver & Battery Co...               TOOLS   
10396                                     Piano Tiles 2™                GAME   
10777                                       Temple Run 2                GAME   
10822                                             Google               TOOLS   
10359  Truecaller: Caller ID, SMS spam blocking & Dialer       COMMUNICATION   
10711                         SHAREit - Transfer & Share               TOOLS   
10389   Sniper 3D Gun Shooter: Free Shooting Games - FPS                GAME   
10676                                   Farm Heroes Saga                GAME   
10576   PicsArt Photo Studio: Collage Maker & Pic Editor         PHOTOGRAPHY   
10461  PhotoGrid: Video & Pic Collage Maker, Photo Ed...         PHOTOGRAPHY   
10502   GO Launcher - 3D parallax Themes & HD Wallpap

If you look at the number of reviews, you can find the most popular apps on the Android App Store. These include the usual suspects: Facebook, WhatsApp, Instagram etc. What’s also notable is that the list of the top 50 most reviewed apps does not include a single paid app! 🤔

# Plotly Pie and Donut Charts - Visualise Categorical Data: Content Ratings

In [ ]:
ratings = df_apps_clean['Content_Rating'].value_counts()
ratings

Content_Rating
Everyone           6621
Teen                912
Mature 17+          357
Everyone 10+        305
Adults only 18+       3
Unrated               1
Name: count, dtype: int64

In [ ]:
fig = px.pie(labels=ratings.index, values=ratings.values, title="Content Rating", names=ratings.index, hole=0.6)

fig.update_traces(textposition='inside', textinfo='percent+label')

fig.show()

# Numeric Type Conversion: Examine the Number of Installs

How many apps had over 1 billion (that's right - BILLION) installations? How many apps just had a single install?

Check the datatype of the Installs column.

Count the number of apps at each level of installations.

Convert the number of installations (the Installs column) to a numeric data type. Hint: this is a 2-step process. You'll have make sure you remove non-numeric characters first.

In [ ]:
df_apps_clean.Installs.describe()

count          8199
unique           19
top       1,000,000
freq           1417
Name: Installs, dtype: object

In [ ]:
df_apps_clean.Installs = df_apps_clean.Installs.astype(str).str.replace(',', "")
df_apps_clean.Installs = pd.to_numeric(df_apps_clean.Installs)
df_apps_clean[['App', 'Installs']].groupby('Installs').count()

App
Installs        
1              3
5              9
10            69
50            56
100          303
500          199
1000         698
5000         425
10000        988
50000        457
100000      1096
500000       504
1000000     1417
5000000      607
10000000     933
50000000     202
100000000    189
500000000     24
1000000000    20

# Find the Most Expensive Apps, Filter out the Junk, and Calculate a (ballpark) Sales Revenue Estimate

Let's examine the Price column more closely.

Convert the price column to numeric data. Then investigate the top 20 most expensive apps in the dataset.

Remove all apps that cost more than $250 from the `df_apps_clean` DataFrame.

Add a column called 'Revenue_Estimate' to the DataFrame. This column should hold the price of the app times the number of installs. What are the top 10 highest grossing paid apps according to this estimate? Out of the top 10 highest grossing paid apps, how many are games?


In [ ]:
df_apps_clean.Price = df_apps_clean.Price.astype(str).str.replace('$', "")
df_apps_clean.Price = pd.to_numeric(df_apps_clean.Price)
df_apps_clean.sort_values(by="Price", ascending=False).head(20)

App   Category  Rating  Reviews  Size_MBs  \
3946        I'm Rich - Trump Edition  LIFESTYLE    3.60      275      7.30   
2394                      I am Rich!    FINANCE    3.80       93     22.00   
2461              I AM RICH PRO PLUS    FINANCE    4.00       36     41.00   
1331          most expensive app (H)     FAMILY    4.30        6      1.50   
3145              I am rich(premium)    FINANCE    3.50      472      0.94   
2775                   I Am Rich Pro     FAMILY    4.40      201      2.70   
5765                       I am rich  LIFESTYLE    3.80     3547      1.80   
3221                  I am Rich Plus     FAMILY    4.00      856      8.70   
3554                      💎 I'm rich  LIFESTYLE    3.80      718     26.00   
3114                       I am Rich    FINANCE    4.30      180      3.80   
4606               I Am Rich Premium    FINANCE    4.10     1867      4.70   
1946  I am rich (Most expensive app)    FINANCE    4.10      129      2.70   
3897                       I Am Rich     FAMILY    3.60      217      4.90   
2193             I am extremely Rich  LIFESTYLE    2.90       41      2.90   
3856                   I am rich VIP  LIFESTYLE    3.80      411      2.60   
2281       Vargo Anesthesia Mega App    MEDICAL    4.60       92     32.00   
1407                    LTC AS Legal    MEDICAL    4.00        6      1.30   
2629                I am Rich Person  LIFESTYLE    4.20      134      1.80   
2481         A Manual of Acupuncture    MEDICAL    3.50      214     68.00   
2207                        EMT PASS    MEDICAL    3.40       51      2.40   

      Installs  Type  Price Content_Rating         Genres  
3946     10000  Paid 400.00       Everyone      Lifestyle  
2394      1000  Paid 399.99       Everyone        Finance  
2461      1000  Paid 399.99       Everyone        Finance  
1331       100  Paid 399.99       Everyone  Entertainment  
3145      5000  Paid 399.99       Everyone        Finance  
2775      5000  Paid 399.99       Everyone  Entertainment  
5765    100000  Paid 399.99       Everyone      Lifestyle  
3221     10000  Paid 399.99       Everyone  Entertainment  
3554     10000  Paid 399.99       Everyone      Lifestyle  
3114      5000  Paid 399.99       Everyone        Finance  
4606     50000  Paid 399.99       Everyone        Finance  
1946      1000  Paid 399.99           Teen        Finance  
3897     10000  Paid 389.99       Everyone  Entertainment  
2193      1000  Paid 379.99       Everyone      Lifestyle  
3856     10000  Paid 299.99       Everyone      Lifestyle  
2281      1000  Paid  79.99       Everyone        Medical  
1407       100  Paid  39.99       Everyone        Medical  
2629      1000  Paid  37.99       Everyone      Lifestyle  
2481      1000  Paid  33.99       Everyone        Medical  
2207      1000  Paid  29.99       Everyone        Medical

There are 15 I am Rich Apps in the Google Play Store apparently. They all cost 300 or more, which is the main point of the app. The story goes that in 2008, Armin Heinrich released the very first I am Rich app in the iOS App Store for 999.90. The app does absolutely nothing. It just displays the picture of a gemstone and can be used to prove to your friends how rich you are. Armin actually made a total of 7 sales before the app was hastily removed by Apple. Nonetheless, it inspired a bunch of copycats on the Android App Store, but if you search today, you’ll find all of these apps have disappeared as well. The high installation numbers are likely gamed by making the app was available for free at some point to get reviews and appear more legitimate.

### The most expensive apps sub $250

In [ ]:
df_apps_clean = df_apps_clean[df_apps_clean.Price < 250]
df_apps_clean.sort_values(by="Price", ascending=False).head(20)

App     Category  Rating  \
2281                          Vargo Anesthesia Mega App      MEDICAL    4.60   
1407                                       LTC AS Legal      MEDICAL    4.00   
2629                                   I am Rich Person    LIFESTYLE    4.20   
2481                            A Manual of Acupuncture      MEDICAL    3.50   
504                           AP Art History Flashcards       FAMILY    5.00   
2463                                 PTA Content Master      MEDICAL    4.20   
2207                                           EMT PASS      MEDICAL    3.40   
4264                            Golfshot Plus: Golf GPS       SPORTS    4.10   
4772   Human Anatomy Atlas 2018: Complete 3D Human Body      MEDICAL    4.50   
3241  Muscle Premium - Human Anatomy, Kinesiology, B...      MEDICAL    4.20   
2293                               Hospitalist Handbook      MEDICAL    4.80   
4470                                  DRAGON QUEST VIII       FAMILY    4.50   
526                          USMLE Step 2 CK Flashcards       FAMILY    5.00   
2119                                         NewTek NDI  PHOTOGRAPHY    3.50   
2473                            boattheory.ch Full 2018       FAMILY    4.70   
4090                             I am Rich Premium Plus      FINANCE    4.60   
3778                            The World Ends With You         GAME    4.60   
1508                                 SkyTest BU/GU Lite     BUSINESS    2.90   
3439                            Trine 2: Complete Story         GAME    3.80   
2603                         2017 EMRA Antibiotic Guide      MEDICAL    4.40   

      Reviews  Size_MBs  Installs  Type  Price Content_Rating        Genres  
2281       92     32.00      1000  Paid  79.99       Everyone       Medical  
1407        6      1.30       100  Paid  39.99       Everyone       Medical  
2629      134      1.80      1000  Paid  37.99       Everyone     Lifestyle  
2481      214     68.00      1000  Paid  33.99       Everyone       Medical  
504         1     96.00        10  Paid  29.99     Mature 17+     Education  
2463       64     41.00      1000  Paid  29.99       Everyone       Medical  
2207       51      2.40      1000  Paid  29.99       Everyone       Medical  
4264     3387     25.00     50000  Paid  29.99       Everyone        Sports  
4772     2921     25.00    100000  Paid  24.99       Everyone       Medical  
3241      168     25.00     10000  Paid  24.99       Everyone       Medical  
2293       12     18.00      1000  Paid  19.99       Everyone       Medical  
4470     7812     27.00     50000  Paid  19.99   Everyone 10+  Role Playing  
526         1     40.00        10  Paid  19.99       Everyone     Education  
2119       77      1.20      1000  Paid  19.99       Everyone   Photography  
2473       54     50.00      1000  Paid  19.40       Everyone     Education  
4090      459      2.00     10000  Paid  18.99       Everyone       Finance  
3778     4108     13.00     10000  Paid  17.99   Everyone 10+        Arcade  
1508       28     20.00       500  Paid  17.99       Everyone      Business  
3439      252     11.00     10000  Paid  16.99           Teen        Action  
2603       12      3.80      1000  Paid  16.99       Everyone       Medical

When we look at the top 5 apps now, we see that 4 out of 5 are medical apps.

### Highest Grossing Paid Apps (ballpark estimate)

In [ ]:
df_apps_clean['Revenue_Estimate'] = df_apps_clean.Installs.mul(df_apps_clean.Price)
df_apps_clean.sort_values(by='Revenue_Estimate', ascending=False)[:10]

App     Category  Rating  Reviews  Size_MBs  \
9220                      Minecraft       FAMILY    4.50  2376564     19.00   
8825                  Hitman Sniper         GAME    4.60   408292     29.00   
7151  Grand Theft Auto: San Andreas         GAME    4.40   348962     26.00   
7477            Facetune - For Free  PHOTOGRAPHY    4.40    49553     48.00   
7977        Sleep as Android Unlock    LIFESTYLE    4.50    23966      0.85   
6594            DraStic DS Emulator         GAME    4.60    87766     12.00   
6082                   Weather Live      WEATHER    4.50    76593      4.75   
6746     Card Wars - Adventure Time       FAMILY    4.30   129603     23.00   
7633        Five Nights at Freddy's         GAME    4.60   100805     50.00   
7954                    Bloons TD 5       FAMILY    4.60   190086     94.00   

      Installs  Type  Price Content_Rating                     Genres  \
9220  10000000  Paid   6.99   Everyone 10+  Arcade;Action & Adventure   
8825  10000000  Paid   0.99     Mature 17+                     Action   
7151   1000000  Paid   6.99     Mature 17+                     Action   
7477   1000000  Paid   5.99       Everyone                Photography   
7977   1000000  Paid   5.99       Everyone                  Lifestyle   
6594   1000000  Paid   4.99       Everyone                     Action   
6082    500000  Paid   5.99       Everyone                    Weather   
6746   1000000  Paid   2.99   Everyone 10+    Card;Action & Adventure   
7633   1000000  Paid   2.99           Teen                     Action   
7954   1000000  Paid   2.99       Everyone                   Strategy   

      Revenue_Estimate  
9220     69,900,000.00  
8825      9,900,000.00  
7151      6,990,000.00  
7477      5,990,000.00  
7977      5,990,000.00  
6594      4,990,000.00  
6082      2,995,000.00  
6746      2,990,000.00  
7633      2,990,000.00  
7954      2,990,000.00

This generously assumes of course that all the installs would have been made at the listed price, which is unlikely, as there are always promotions and free give-aways on the App Stores.

The top spot of the highest-grossing paid app goes to … Minecraft at close to $70 million. It’s quite interesting that Minecraft (along with Bloons and Card Wars) is actually listed in the Family category rather than in the Game category. If we include these titles, we see that 7 out the top 10 highest-grossing apps are games. The Google Play Store seems to be quite flexible with its category labels.

# Plotly Bar Charts & Scatter Plots: Analysing App Categories

In [ ]:
df_apps_clean.Category.nunique()

33

### Vertical Bar Chart - Highest Competition (Number of Apps)

In [ ]:
top_10_category = df_apps_clean.Category.value_counts()[:10]
top_10_category

Category
FAMILY             1606
GAME                910
TOOLS               719
PRODUCTIVITY        301
PERSONALIZATION     298
LIFESTYLE           297
FINANCE             296
MEDICAL             292
PHOTOGRAPHY         263
BUSINESS            262
Name: count, dtype: int64

In [ ]:
bar = px.bar(x=top_10_category.index,
             y=top_10_category.values,
             title="Category Popularity")

bar.update_layout(xaxis_title="Category", yaxis_title="Number of Apps")
bar.show()

Based on the number of apps, the Family and Game categories are the most competitive. Releasing yet another app into these categories will make it hard to get noticed.

### Horizontal Bar Chart - Most Popular Categories (Highest Downloads)

In [ ]:
category_installs = df_apps_clean[['Category', 'Installs']].groupby('Category').sum()
category_installs.sort_values(by='Installs', ascending=True, inplace=True)

In [ ]:
h_bar = px.bar(x=category_installs.Installs,
               y=category_installs.index,
               orientation="h",
               title="Category Popularity")

h_bar.update_layout(xaxis_title="Total Installs", yaxis_title="Category")

h_bar.show()

### Category Concentration - Downloads vs. Competition

* First, create a DataFrame that has the number of apps in one column and the number of installs in another

* Then use the [plotly express examples from the documentation](https://plotly.com/python/line-and-scatter/) alongside the [.scatter() API reference](https://plotly.com/python-api-reference/generated/plotly.express.scatter.html)to create scatter plot that looks like this.

In [ ]:
df_category_concentration = df_apps_clean[['Category', 'App']].groupby('Category').count()
df_category_concentration['Installs'] = df_apps_clean[['Category', 'Installs']].groupby('Category').sum()['Installs']
df_category_concentration.head()

App    Installs
Category                            
ART_AND_DESIGN        61   114233100
AUTO_AND_VEHICLES     73    53129800
BEAUTY                42    26916200
BOOKS_AND_REFERENCE  169  1665791655
BUSINESS             262   692018120

In [ ]:
fig = px.scatter(df_category_concentration,
                 x='App',
                 y='Installs',
                 title='Category Concentration',
                 size='App',
                 hover_name=df_category_concentration.index,
                 color='Installs')

fig.update_layout(xaxis_title='Number of Apps', yaxis_title='Installs')

fig.show()

What we see is that the categories like Family, Tools, and Game have many different apps sharing a high number of downloads. But for the categories like video players and entertainment, all the downloads are concentrated in very few apps.

# Extracting Nested Data from a Column

How many different types of genres are there? Can an app belong to more than one genre? Check what happens when you use .value_counts() on a column with nested values? See if you can work around this problem by using the .split() function and the DataFrame's [.stack() method](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.stack.html).


In [ ]:
df_apps_clean.head()

App Category  Rating  Reviews  \
21                              KBA-EZ Health Guide  MEDICAL    5.00        4   
28                                         Ra Ga Ba     GAME    5.00        2   
47                                          Mu.F.O.     GAME    5.00        2   
82                                 Brick Breaker BR     GAME    5.00        7   
99  Anatomy & Physiology Vocabulary Exam Review App  MEDICAL    5.00        1   

    Size_MBs  Installs  Type  Price Content_Rating   Genres  Revenue_Estimate  
21     25.00         1  Free   0.00       Everyone  Medical              0.00  
28     20.00         1  Paid   1.49       Everyone   Arcade              1.49  
47     16.00         1  Paid   0.99       Everyone   Arcade              0.99  
82     19.00         5  Free   0.00       Everyone   Arcade              0.00  
99      4.60         5  Free   0.00       Everyone  Medical              0.00

In [ ]:
stack = df_apps_clean.Genres.str.split(';', expand=True).stack()
num_genres = stack.value_counts()
print(f'Number of genres: {len(num_genres)}')

Number of genres: 53


# Colour Scales in Plotly Charts - Competition in Genres

In [ ]:
top_15_genres = num_genres[:15]
fig = px.bar(x=top_15_genres.index, y=top_15_genres.values,
             title='Top Genres',
             color=top_15_genres.values,
             color_continuous_scale="Viridis")

fig.update_layout(xaxis_title='Genre', yaxis_title='Number of Apps', coloraxis_showscale=False)
fig.show()

# Grouped Bar Charts: Free vs. Paid Apps per Category

In [ ]:
df_free_vs_paid = df_apps_clean[['App','Category','Type']].groupby(['Category','Type'], as_index=False).count()
df_free_vs_paid.head()

Category  Type  App
0     ART_AND_DESIGN  Free   58
1     ART_AND_DESIGN  Paid    3
2  AUTO_AND_VEHICLES  Free   72
3  AUTO_AND_VEHICLES  Paid    1
4             BEAUTY  Free   42

In [ ]:
g_bar = px.bar(df_free_vs_paid,
               x='Category',
               y='App',
               title='Free vs Paid Apps by Category',
               color='Type',
               barmode='group')

g_bar.update_layout(xaxis_title='Category',
                    yaxis_title='Number of Apps',
                    xaxis={'categoryorder':'total descending'},
                    yaxis=dict(type='log'))

g_bar.show()

What we see is that while there are very few paid apps on the Google Play Store, some categories have relatively more paid apps than others, including Personalization, Medical and Weather. So, depending on the category you are targeting, it might make sense to release a paid-for app.

# Plotly Box Plots: Lost Downloads for Paid Apps


In [88]:
box = px.box(df_apps_clean,
             y='Installs',
             x='Type',
             color='Type',
             notched=True,
             points='all',
             title='How Many Downloads are Paid Apps Giving Up?')

box.update_layout(yaxis=dict(type='log'))

box.show()

From the hover text in the chart, we see that the median number of downloads for free apps is 500,000, while the median number of downloads for paid apps is around 5,000! This is massively lower.

# Plotly Box Plots: Revenue by App Category


In [92]:
df_paid_apps = df_apps_clean[df_apps_clean['Type'] == 'Paid']
box = px.box(df_paid_apps,
             x='Category',
             y='Revenue_Estimate',
             title='How Much Can Paid Apps Earn?')

box.update_layout(xaxis_title='Category',
                  yaxis_title='Paid App Ballpark Revenue',
                  xaxis={'categoryorder':'min ascending'},
                  yaxis=dict(type='log'))


box.show()

If an Android app costs 30,000 to develop, then the average app in very few categories would cover that development cost. The median paid photography app earned about 20,000. Many more app’s revenues were even lower - meaning they would need other sources of revenue like advertising or in-app purchases to make up for their development costs. However, certain app categories seem to contain a large number of outliers that have much higher (estimated) revenue - for example in Medical, Personalisation, Tools, Game, and Family.

# How Much Can You Charge? Examine Paid App Pricing Strategies by Category

What is the median price price for a paid app? Then compare pricing by category by creating another box plot. But this time examine the prices (instead of the revenue estimates) of the paid apps. I recommend using `{categoryorder':'max descending'}` to sort the categories.

In [94]:
df_paid_apps = df_apps_clean[df_apps_clean['Type'] == 'Paid']
box = px.box(df_paid_apps,
             x='Category',
             y='Price',
             title='Price per Category')

box.update_layout(xaxis_title='Category',
                  yaxis_title='Paid App Price',
                  xaxis={'categoryorder':'max descending'},
                  yaxis=dict(type='log'))


box.show()

The median price for an Android app is 2.99 USD.

However, some categories have higher median prices than others. This time we see that Medical apps have the most expensive apps as well as a median price of 5.49 USD. In contrast, Personalisation apps are quite cheap on average at $1.49. Other categories which higher median prices are Business (4.99 USD) and Dating (6.99 USD). It seems like customers who shop in these categories are not so concerned about paying a bit extra for their apps.